### DISCUSSION TITLE
Database Setup

### DISCUSSION THREAD

**Discussion by marinjurisich** [marinjurisich/TimeKeeper#21] [Announcements]
Pozz, danas cemo nauciti kako postaviti bazu za ovaj projekt 😃.

1. Prvo treba instalirati sql server. SQL server je besplatan za developere, download je na ovom linku: https://www.microsoft.com/en/sql-server/sql-server-downloads.

2. Nakon toga instalirajte Sql Server Management System (SSMS) s ovog linka: https://learn.microsoft.com/en-us/sql/ssms/download-sql-server-management-studio-ssms?view=sql-server-ver16.

3. U SSMS-u otvorite novi sql server, po mogućnosti naziva ".\\sqlexpress". Ako ga nazovete drugacije i to je u redu.

4. Provjerite da vam se connection string u projektu poklapa s nazivom servera na sljedeći način:
 Naziv servera izgleda ovako i u ovom slučaju to je ".":
![a8c5c226-9b9a-4540-bc01-cc02103c73d4](https://user-images.githubusercontent.com/72669095/227002981-69a5d9b9-70af-4c44-b317-f4f2aefd076a.jpg)

Connection string se nalazi u `appsettings.json` datoteci u TimeKeeper projektu i izgleda ovako:
![1892fee5-528b-4bbf-b5a9-cad496628253](https://user-images.githubusercontent.com/72669095/227003291-daf6ae45-6824-45ff-860f-15c487584af6.jpg)
Ako je u SSMS-u naziv servera ".", tako mora biti i u projektu. Ako budete morali mijenjati naziv servera u connection stringu, nemojte pushati tu promjenu na Github, budem napravio da se taj file exclude-a iz git changes-a nakon što svi preuzmemo trenutnu verziju.

5. Ako ste dosad sve dobro odradili trebate prijeći na posljednji korak - izvršavanje migracija! Otvorite Visual Studio 2022, i na vrhu pod View nađite Developer Powershell konzolu. Otvorite ju i u nju upišite sljedeće:
`dotnet ef database update`
Ako dobijete error instalirajte Entity Framework koristeći naredbu 
`dotnet tool install --global dotnet-ef`
nakon čega opet treba pozvati `dotnet ef database update`.
To možete vidjeti i na sljedećoj slici:
![26448e28-f2bc-4c4b-804a-4c5abf446842](https://user-images.githubusercontent.com/72669095/227004499-5b10f3bd-16be-419a-97bb-a5b7e697bde9.jpg)
Ako sve izgleda kao na gornjoj slici, trebali biste refreshati Sql Server u SSMS-u (desni klik na server u Object Explorer -> refresh), nakon čega bi vam se trebala pojaviti TimeKeeper baza podataka sa svim potrebnim tablicama i setom "dummy" podataka. Enjoy!

**Comment by shtef21** [comment]
Konačno mi je proradila baza na Windowsu :D
Lijepo si ovo napisao, svaka čast

**Comment by shtef21** [comment]
Pronašao sam službene upute za Linux, pa stavljam.

# Install SQL Server on Ubuntu 20.04

[Source](https://learn.microsoft.com/en-us/sql/linux/quickstart-install-connect-ubuntu?view=sql-server-ver16)

## Install SQL Server

### 1. Import the public repository GPG keys:

```bash
wget -qO- https://packages.microsoft.com/keys/microsoft.asc | sudo tee /etc/apt/trusted.gpg.d/microsoft.asc
```

### 2. Register the SQL Server Ubuntu repository:

```bash
sudo add-apt-repository "$(wget -qO- https://packages.microsoft.com/config/ubuntu/20.04/mssql-server-2022.list)"
```

### 3. Run the following commands to install SQL Server:

```bash
sudo apt-get update && sudo apt-get install -y mssql-server
```

### 4. After the package installation finishes, run `mssql-conf setup` and follow the prompts to set the SA password and choose your edition. As a reminder, the following SQL Server editions are freely licensed: Evaluation, Developer, and Express.

```bash
sudo /opt/mssql/bin/mssql-conf setup
```

Remember to specify a strong password for the SA account. You need a minimum length 8 characters, including uppercase and lowercase letters, base-10 digits and/or non-alphanumeric symbols.

### 5. Once the configuration is done, verify that the service is running:

```bash
systemctl status mssql-server --no-pager
```

### 6. If you plan to connect remotely, you might also need to open the SQL Server TCP port (default 1433) on your firewall.

### At this point, SQL Server is running on your Ubuntu machine and is ready to use!



##  Install the SQL Server command-line tools

Use the following steps to install the `mssql-tools` on Ubuntu. If **curl** isn't installed, you can run this code:

```bash
sudo apt-get update && sudo apt install curl
```

### 1. Import the public repository GPG keys:

```bash
curl https://packages.microsoft.com/keys/microsoft.asc | sudo tee /etc/apt/trusted.gpg.d/microsoft.asc
```

### 2. Register the Ubuntu repository:

```bash
curl https://packages.microsoft.com/config/ubuntu/20.04/prod.list | sudo tee /etc/apt/sources.list.d/msprod.list
```

### 3. Update the sources list and run the installation command with the `unixODBC` developer package.

```bash
sudo apt-get update && sudo apt-get install mssql-tools unixodbc-dev
```

You can update to the latest version of mssql-tools using the following commands:

```bash
sudo apt-get update && sudo apt-get install mssql-tools
```

### 4. Add `sqlcmd` and `bcp` to PATH to make it available in terminal by default.

Add the following code to your `~/.bashrc` file with any GUI / CLI text editor:

```bash
export PATH="$PATH:/opt/mssql-tools/bin"
```

## Connect locally

```bash
sqlcmd -S localhost -U sa -P '<YourPassword>'
```

Where:
- `-S` is SQL Server name
- `-U` is username
- `-P` is password

If successful, you should get to a `sqlcmd` command prompt: `1>`

Example usage:

```bash
1>  CREATE DATABASE TimeKeeper
2>  GO
```
It is mandatory to write `GO` for the commands before it to execute.
It seems that writing commands in `sqlcmd` resembles writing a transaction.


## Rename SQL Server

Enter the SQL CLI:
```bash
sqlcmd -S localhost -U sa -P '<YourPassword>'
```

Follow these instructions.
```bash
  # Find out current server name
1> SELECT @@SERVERNAME
2> GO

  # Drop current server
3> sp_dropserver SERVER_NAME
4> GO

  # Create new server
5> sp_addserver NEW_SERVER_NAME, local
```
Note: commands written with # should not be written.

Lastly, restart SQL Server service:
```bash
systemctl restart mssql-server
```
